src: add support for nogc types via `NogcEnv`

KevinEady opened the pull request:
- Introduce `Napi::NogcEnv` class, taking the place of `Napi::Env` in finalizer callbacks.
- Introduce `void Napi::NogcEnv::AddPostFinalizer()` and overloads.
- Modify tests to support above changes

TODO:
- [ ] docs
- [ ] discuss implementation, namings, etc

KevinEady commented:
@vmoroz your comment on removing const ref on `std::nullptr_t` has been addressed, and @mhdawson your comment on reference deletion has been addressed, both in [2a689b6](https://github.com/nodejs/node-addon-api/pull/1514/commits/2a689b6bde71b731cb29a2a0b5aec244a2fc59ca)

vmoroz commented on .github/workflows/ci.yml:
```diff
@@ -12,8 +12,12 @@ jobs:
+      fail-fast: false
       matrix:
-        node-version: [ 18.x, 20.x, 21.x, 22.x ]
+        api_version:
```
Should we add checks for the standard/experimental to ci-win.yml too?

KevinEady replied:
Good thought... I pretty much copied Gabriel's CI file. I'll comment on his PR to verify, since mine would end up being overwritten when i rebase to master after his is merged.

KevinEady commented:
Hi team,

Using some SFINAE magic, I was able to determine at compile time how to associate the finalizer -- either directly or with `node_api_post_finalizer`. This would address #1367 as implemented.

This means no changes were needed to any existing tests, and I created a new test in `finalizer_order` to ensure finalizers are being called correctly.

PTAL! @gabrielschulhof @vmoroz @legendecas @mhdawson

gabrielschulhof commented on test/finalizer_order.js:
```diff
@@ -0,0 +1,98 @@
+      if (isExperimental) {
+        assert.strictEqual(binding.finalizer_order.Test.isNogcFinalizerCalled, true, 'Expected nogc finalizer to be called [before ticking]');
+        assert.strictEqual(binding.finalizer_order.Test.isGcFinalizerCalled, false, 'Expected gc finalizer to not be called [before ticking]');
+        assert.strictEqual(isCallbackCalled, false, 'Expected callback not be called [before ticking]');
```
```suggestion
        assert.strictEqual(isCallbackCalled, false, 'Expected callback to not be called [before ticking]');
```

gabrielschulhof reviewed (approved):
I think the Gc vs. NoGc nomenclature is a bit confusing. Unfortunately, we established this in core. I think maybe we should go with `DuringGc` and `OutsideGc`. For env, we'd have `DuringGcEnv` and `Env`, since `Env` is already established, but for the finalizer type names I think we can go with `DuringGc` and `OutsideGc` for clarity.

gabrielschulhof commented on napi.h:
```diff
@@ -2415,6 +2448,7 @@ class ObjectWrap : public InstanceWrap<T>, public Reference<Object> {
       napi_property_attributes attributes = napi_default);
   static Napi::Value OnCalledAsFunction(const Napi::CallbackInfo& callbackInfo);
   virtual void Finalize(Napi::Env env);
+  virtual void Finalize(NogcEnv env);
```
```suggestion
  virtual void Finalize(Napi::NogcEnv env);
```

gabrielschulhof commented:
I tested this with `Napi::External` and it works great!

```JS
const { makeSync, makeAsync } = require('bindings')('test_sync_fini')

const doAsync = process.argv[2] === 'async'

;(function test(iter = 0) {
  for (let i = 0; i < 1000000; i++) {
    const id = `${iter}:${i}`
    if (doAsync) {
      makeAsync(id)
    } else {
      makeSync(id)
    }
  }
  setTimeout(test, 0, iter + 1)
})()
```

Here's a video of the output. Guess which column is async 🙂 

https://github.com/nodejs/node-addon-api/assets/976081/1eb30993-926c-4233-8495-93953f112ace

gabrielschulhof commented:
We should rename `AddPostFinalizer` to just `PostFinalizer`.

KevinEady commented:
@gabrielschulhof ,

>  For env, we'd have DuringGcEnv and Env, since Env is already established, ...

One could argue that being a Node-API wrapper, the naming in this library should be as consistent with Node-API as possible. The (almost) 1-to-1 mapping helps link up with other documentation (eg. referencing the [Node-API](https://nodejs.org/api/n-api.html) docs), or translating between Node-API C code and node-addon-api C++ code.

I do agree though, the name `DuringGcEnv` better describes the restriction ("during GC") on how the environment can be used, but not sure if we should stray from the Node-API names.

---

 > ... but for the finalizer type names I think we can go with `DuringGc` and `OutsideGc` for clarity.

By "finalizer type names" do you mean changing the `Finalizer` in places eg.:
```cpp
template <typename Finalizer>
static External New(napi_env env, T* data, Finalizer finalizeCallback);
```

If so, I don't think we should change this name since the finalizer passed can accept either a `DuringGcEnv` or `Env`. A `Finalizer` is a callback ran either during the GC cycle (if using `DuringGcEnv`) or on a future tick after the GC cycle completes (if using `Env`). This means the concept of "finalization" is dependent on your callback signature, and would be documented as such.

---

> We should rename AddPostFinalizer to just PostFinalizer.

Makes sense, esp. considering what I said previously.

---

While on the subject of documentation: we don't have any specific entry for "finalization", as a finalizer callback is always documented on each individual object or function, eg:

- `Napi::External`:
https://github.com/nodejs/node-addon-api/blob/bf49519a4ce08ee5320327c9a0199cd89d5b87b3/doc/external.md?plain=1#L7
- `Napi::Buffer::New`:
https://github.com/nodejs/node-addon-api/blob/bf49519a4ce08ee5320327c9a0199cd89d5b87b3/doc/buffer.md?plain=1#L73-L75
- and others

I propose we create a new top-level documentation page for finalization describing the above, and can have each use of finalizers in the documentation link to this page.

KevinEady commented on napi.h:
```diff
@@ -312,30 +313,31 @@ using MaybeOrValue = T;
- private:
-  napi_env _env;
+class NogcEnv {
+ protected:
```
Should we make this `private` and have `Env` be a friend class? With it being `protected`, the default finalizers (as well as `_env`) would be exposed in any user-defined classes that extend `NogcEnv`.

KevinEady commented:
I've added some WIP documentation on the top-level finalization page: https://github.com/nodejs/node-addon-api/blob/6591a3609c2b94f22956b72736e06b86586556cf/doc/finalization.md

KevinEady commented:
@mhdawson @gabrielschulhof as discussed in the 5 July Node-API meeting, I updated the docs to use the concept of a "basic finalizer" as a special categorization of finalizers, and removing the concept of "[a]synchronous". PTAL: https://github.com/nodejs/node-addon-api/blob/7d7b8211552216657494144d0417bf57b690871e/doc/finalization.md 

Once we finalize (hah) this verbiage, I will update any associated type names in the headers + tests, and update the other docs that have any finalization features.

gabrielschulhof commented:
@KevinEady the doc LGTM 👍

legendecas commented:
(this is still a draft PR)

mhdawson commented:
doc LGTM with a few suggestions:

1) I would remove 

NOTE: Optimizations via basic finalizers will only occur if using NAPI_EXPERIMENTAL and the NODE_API_EXPERIMENTAL_NOGC_ENV_OPT_OUT define flag has not been set. Otherwise, the engine will not differentiate between basic and (extended) finalizers.

there is never any promise that optimization will take place so I think we can just remove that.

2) In terms of 

In addition to passing finalizers to Napi::Externals and other Node-API constructs, use Napi::BasicEnv::PostFinalize(Napi::BasicEnv, Finalizer) to schedule a callback to run outside of the garbage collector finalization.

can we explain without promising it will run outside of gc collection?  Maybe something like "to maximize the likelyhood that the finalization not being tied to garbage collection finalization" ...

KevinEady commented:
@legendecas the overall C++ work/implementation is finished, but the PR still needs some more work (markdown docs, test docs, ...) 

@mhdawson 

> 1. I would remove
> 
> NOTE: Optimizations via basic finalizers will only occur if using NAPI_EXPERIMENTAL and the NODE_API_EXPERIMENTAL_NOGC_ENV_OPT_OUT define flag has not been set. Otherwise, the engine will not differentiate between basic and (extended) finalizers.
> 
> there is never any promise that optimization will take place so I think we can just remove that.

With this, I was trying to convey that using basic finalizers only changes functionality in experimental mode. I could see someone using basic finalizers without experimental mode, running into https://github.com/nodejs/node-addon-api/issues/1213 and being a bit confused. Maybe we can use some other verbiage?

> **NOTE**: Node-API will only differentiate between basic and (extended) finalizers when using `NAPI_EXPERIMENTAL` and the `NODE_API_EXPERIMENTAL_NOGC_ENV_OPT_OUT` define flag has _not_ been set. Otherwise, basic finalizers execute in the same manner as (extended) finalizers.

What do we think of that?

----


> 2. In terms of
> 
> In addition to passing finalizers to Napi::Externals and other Node-API constructs, use Napi::BasicEnv::PostFinalize(Napi::BasicEnv, Finalizer) to schedule a callback to run outside of the garbage collector finalization.
> 
> can we explain without promising it will run outside of gc collection?

Well, I took inspiration from the [Node-API docs for `node_api_post_finalizer`](https://nodejs.org/api/n-api.html#node_api_post_finalizer):

> node_api_post_finalizer helps to work around this limitation by allowing the add-on to defer calls to such Node-APIs to a point in time outside of the GC finalization.

Also, when you say:

> "to maximize the likelyhood that the finalization not being tied to garbage collection finalization"

The API is intended to _guarantee_ the callback does not run inside GC, so saying "maximize the likelihood" is not completely accurate, as it 100% will not run inside GC finalization. 

Since this API strictly deals with callbacks + GC finalization order, I think it's a good thing to keep.

KevinEady commented:
In Node API meeting 19 July, we discussed:

> 1. I would remove
> 
> NOTE: Optimizations via basic finalizers will only occur if using NAPI_EXPERIMENTAL and the NODE_API_EXPERIMENTAL_NOGC_ENV_OPT_OUT define flag has not been set. Otherwise, the engine will not differentiate between basic and (extended) finalizers.
> 
> there is never any promise that optimization will take place so I think we can just remove that.

Yes, remove this.

----

> 2. In terms of
> 
> In addition to passing finalizers to Napi::Externals and other Node-API constructs, use Napi::BasicEnv::PostFinalize(Napi::BasicEnv, Finalizer) to schedule a callback to run outside of the garbage collector finalization.
> 
> can we explain without promising it will run outside of gc collection?

The name `node_api_post_finalizer` can be made in the context of "running (non-basic) code after (**_post_**) a **_finalizer_**" , so maybe we can change the documentation in core (and also here) if needed, but the current wording is fine.

----

Additionally, we discussed adding an _opt-in_ flag that requires -- at compilation time -- all finalizers are of the basic type, under the realization that:
- if using the C-based Node-API, opting into basic finalizers requires _all_ finalizers to be basic, providing devs a way to guarantee all of their finalizers would be run in the "maybe optimized" way and fail compilation if callbacks do not match the current basic finalizer signature.
- if using this wrapper, (currently) there is no guarantee that all finalizers will be of the basic signature, meaning developers do not have an easy way to forcefully get compilation errors for finalizers that are not basic.

I suggest using a define: `NODE_API_REQUIRE_BASIC_FINALIZERS`

KevinEady commented:
Hi @mhdawson @vmoroz @legendecas @gabrielschulhof ,

I believe this PR is ready for review now! 

- Finished renaming "synchronous" finalizer to "basic" finalizer
- Updated docs
- Removed semver major label since it's a feature add and nothing breaking
- Changed the PR title to `feat:` to get release-please stuff correct
 
PTAL! Hopefully we can discuss in this week's Node API team meeting.

---

> Additionally, we discussed adding an _opt-in_ flag that requires -- at compilation time -- all finalizers are of the basic type, ...

I will defer this to a different PR, as this will need to update test infrastructure (to test for compilation failures)

KevinEady commented:
I would like to land https://github.com/nodejs/node-addon-api/pull/1548 and rebase this PR so we can get a Windows + experimental test run.

KevinEady commented:
The Windows experimental CI passed, eg: [test (experimental, 20.x, x64, windows-2019)](https://github.com/nodejs/node-addon-api/actions/runs/10218072154/job/28273303597#step:7:3698)
```
Testing with Node-API Version '2147483647'.
```

@legendecas @mhdawson @gabrielschulhof @vmoroz PTAL! 👍

mhdawson reviewed (approved):
LGTM

legendecas commented on napi.h:
```diff
@@ -187,6 +187,7 @@ namespace NAPI_CPP_CUSTOM_NAMESPACE {
 #endif
 
 // Forward declarations
+class BasicEnv;
```
Is this necessary?

legendecas reviewed (approved):
Good work! Thanks

legendecas commented on doc/finalization.md:
```diff
@@ -0,0 +1,149 @@
+scheduling or executing the callback. For example, V8 does not allow access to
+the engine heap during garbage collection. Restricting finalizers from accessing
+the engine heap allows the callback to execute during garbage collection,
+providing a chance to free native memory on the current tick.
```
I think it would be good to just say "eagerly" being GC-ed.

```suggestion
providing a chance to free native memory eagerly.
```

KevinEady replied:
Addressed in [195ec28](https://github.com/nodejs/node-addon-api/pull/1514/commits/195ec28a7b1dcfe571c6df5f8a2e1bb991126ff1).

KevinEady commented:
Hi @gabrielschulhof , IIRC you mentioned you had an in-progress review. Any movement on that?

gabrielschulhof commented on doc/env.md:
```diff
@@ -1,8 +1,14 @@
 # Env
 
```
```suggestion
Class `Napi::Env` inherits from class [`Napi::BasicEnv`][].

```

and then add

```
[`Napi::BasicEnv`]: ./basic_env.md
```

to the bottom of the file.

KevinEady replied:
Addressed in [195ec28](https://github.com/nodejs/node-addon-api/pull/1514/commits/195ec28a7b1dcfe571c6df5f8a2e1bb991126ff1).

gabrielschulhof commented on doc/finalization.md:
```diff
@@ -0,0 +1,149 @@
+parameter. This parameter represents a native callback function that runs in
+response to a garbage collection event. A finalizer is considered a _basic_
+finalizer if the callback only utilizes a certain subset of APIs, which may
+provide optimizations, improved execution, or other benefits.
```
```suggestion
provide more efficient memory management, optimizations, improved execution, or other benefits.
```

I think we should explicitly point out memory management because that's what caused us to go down this path in the first place.

KevinEady replied:
Addressed in [195ec28](https://github.com/nodejs/node-addon-api/pull/1514/commits/195ec28a7b1dcfe571c6df5f8a2e1bb991126ff1).

gabrielschulhof commented on doc/finalization.md:
```diff
@@ -0,0 +1,149 @@
+## Scheduling Finalizers
+
+In addition to passing finalizers to `Napi::External`s and other Node-API
+constructs, use `Napi::BasicEnv::PostFinalize(Napi::BasicEnv, Finalizer)` to
```
```suggestion
constructs, `Napi::BasicEnv::PostFinalize(Napi::BasicEnv, Finalizer)` can be used to
```

We should use passive voice.

KevinEady replied:
Addressed in [195ec28](https://github.com/nodejs/node-addon-api/pull/1514/commits/195ec28a7b1dcfe571c6df5f8a2e1bb991126ff1).

KevinEady commented on doc/basic_env.md:
```diff
@@ -0,0 +1,201 @@
+# BasicEnv
+
+The opaque data structure containing the environment in which the request is
```
Addressed in [195ec28](https://github.com/nodejs/node-addon-api/pull/1514/commits/195ec28a7b1dcfe571c6df5f8a2e1bb991126ff1).

KevinEady commented on doc/basic_env.md:
````diff
@@ -0,0 +1,201 @@
+const char* Napi::Env::GetModuleFileName() const;
+```
+
+Returns a A URL containing the absolute path of the location from which the
````
Addressed in [195ec28](https://github.com/nodejs/node-addon-api/pull/1514/commits/195ec28a7b1dcfe571c6df5f8a2e1bb991126ff1).

KevinEady commented on doc/basic_env.md:
```diff
@@ -0,0 +1,201 @@
+`fini` when the addon instance is unloaded.
+
+Associates a data item stored at `T* data` with the current instance of the
+addon. The item will be passed to the function `fini` which gets called when an
```
Addressed in [195ec28](https://github.com/nodejs/node-addon-api/pull/1514/commits/195ec28a7b1dcfe571c6df5f8a2e1bb991126ff1).

KevinEady commented on doc/basic_env.md:
```diff
@@ -0,0 +1,201 @@
+
+Associates a data item stored at `T* data` with the current instance of the
+addon. The item will be passed to the function `fini` which gets called when an
+instance of the addon is unloaded. This overload accepts an additional hint to
```
Addressed in [195ec28](https://github.com/nodejs/node-addon-api/pull/1514/commits/195ec28a7b1dcfe571c6df5f8a2e1bb991126ff1).

mhdawson commented:
Looks like all comments have been addressed and CI is green. Landing.

KevinEady commented:
Hi @legendecas ,

This PR was merged into main, but the `release-please` action failed: https://github.com/nodejs/node-addon-api/actions/runs/10687066491/job/29623803353 

I think it has to do with permissions?
